chrome/content/zotero/xpcom/storage/storageCache.js

```diff
@@ -0,0 +1,319 @@
+		await Zotero.Promise.all(
+			attachmentItems.map(attachmentItem => this._deleteItemFiles(attachmentItem))
+		);
+		
+		// TODO: There seems to be a lag compared to download file of how long it takes the
+		// item pane to update
+	},
+
+	/**
+	 * Cache the given item attachments
+	 *
+	 * @param items {Array} - List of items that are attachments
+	 *
+	 * @returns {Promise<void>}
+	 */
+	cacheItemAttachments: async function (items) {
+		let failed = [];
+		await Zotero.Promise.all(items.map(async (item) => {
+			let path = item.getFilePath();
+			let fileExists = await OS.File.exists(path);
+
+			// TEMP: If file is queued for download, download first. Starting in 5.0.85, files
+			// modified remotely get marked as SYNC_STATE_FORCE_DOWNLOAD, causing them to get
+			// downloaded at sync time even in download-as-needed mode, but this causes files
+			// modified previously to be downloaded on open.
+			if (fileExists
+				&& (item.attachmentSyncState !== Zotero.Sync.Storage.Local.SYNC_STATE_TO_DOWNLOAD)) {
+				// We are not downloading, but still touch last accessed
+				item.attachmentLastAccessed = Date.now();
+				await item.saveTx({ skipAll: true });
+				return;
+			}
+
+			// Download item
+			let results = await Zotero.Sync.Runner.downloadFile(item);
+			if (!results || !results.localChanges) {
+				failed.push(item);
+				Zotero.debug("Attachment download failed -- skipping");
+			}
+		}));
+		
+		Zotero.debug('Finished caching items');
+
+		if (failed.length) {
+			Zotero.alert(
+				window,
+				'Download Failure',
+				'Some attachments failed to download: ' + failed.join(', ')
+			);
+		}
+	},
+
+	/**
+	 * Identify items with files currently downloaded that we should delete to limit cache
+	 *
+	 * Called when the limit cache preference is turned on or when a new item is downloaded or
+	 * added.
+	 *
+	 * @param forGroups {boolean} - Are we identifying in group libraries or the user library
+	 *
+	 * @returns {Promise<void>}
+	 */
+	identifyItemsToFree: async function (forGroups) {
+		this._freeingCache = true;
+		let promises = [];
+		try {
+			if (forGroups) {
+				Zotero.Libraries.getAll().forEach((library) => {
+					if (library.libraryID !== Zotero.Libraries.userLibraryID) {
+						promises.push(
+							this.identifyItemsToFreeForLibrary(library.libraryID)
+						);
+					}
+				});
+			}
+			else {
+				promises.push(
+					this.identifyItemsToFreeForLibrary(Zotero.Libraries.userLibraryID)
+				);
+			}
+	
+			await Zotero.Promise.all(promises);
+		}
+		catch (e) {
+			throw e;
+		}
+		finally {
+			this._freeingCache = false;
+		}
+	},
+
+	/**
+	 * Identify items with files currently downloaded that we should delete to limit cache
+	 *
+	 * Called when the limit cache preference is turned on or when a new item is downloaded or
+	 * added.
+	 *
+	 * @param libraryID {integer} - Library ID to search for items to free
+	 *
+	 * @returns {Promise<void>}
+	 */
+	identifyItemsToFreeForLibrary: async function (libraryID) {
```

Comment:
`identify` is a little misleading here — this is actually doing the deletion, right?

For testing purposes and general cleanliness, maybe it makes sense to split this into two functions, with one that just returns a specified number of items with files to delete.